pkg/providers/instancetype/instancetype.go

```diff
@@ -242,6 +256,80 @@ func (p *DefaultProvider) UpdateInstanceTypeOfferings(ctx context.Context) error
 	return nil
 }
 
+func (p *DefaultProvider) UpdateInstanceTypeMemoryOverhead(ctx context.Context, kubeClient client.Client) error {
+	nodeClaimList := &karpv1.NodeClaimList{}
+	if err := kubeClient.List(ctx, nodeClaimList); err != nil {
+		return fmt.Errorf("failed to list nodeclaims: %w", err)
+	}
+	nodeToNodeClaims := lo.Associate(nodeClaimList.Items, func(nc karpv1.NodeClaim) (string, karpv1.NodeClaim) {
+		return nc.Status.NodeName, nc
+	})
+
+	nodeClassList := &v1.EC2NodeClassList{}
+	if err := kubeClient.List(ctx, nodeClassList); err != nil {
+		return fmt.Errorf("failed to list nodeclasses: %w", err)
+	}
+	nodeClassMap := lo.Associate(nodeClassList.Items, func(nc v1.EC2NodeClass) (string, v1.EC2NodeClass) {
+		return nc.Name, nc
+	})
+
+	// List only Karpenter registered Nodes
+	nodeList := &corev1.NodeList{}
+	if err := kubeClient.List(ctx, nodeList, &client.ListOptions{
+		LabelSelector: labels.SelectorFromSet(map[string]string{karpv1.NodeRegisteredLabelKey: "true"}),
+	}); err != nil {
+		return fmt.Errorf("failed to list nodes: %w", err)
+	}
+
+	instanceTypeInfoMap := lo.Associate(p.instanceTypesInfo, func(i *ec2.InstanceTypeInfo) (string, *ec2.InstanceTypeInfo) {
+		return *i.InstanceType, i
+	})
+
+	workqueue.ParallelizeUntil(ctx, 100, len(nodeList.Items), func(i int) {
+		node := nodeList.Items[i]
+		nodeClaim, ok := nodeToNodeClaims[node.Name]
+		if !ok {
+			return
+		}
+		nodeClass, ok := nodeClassMap[nodeClaim.Spec.NodeClassRef.Name]
+		if !ok {
+			return
+		}
+
+		// Ensure AMI is current
+		if !lo.ContainsBy(nodeClass.Status.AMIs, func(ami v1.AMI) bool {
+			return ami.ID == nodeClaim.Status.ImageID
+		}) {
+			return
+		}
+
+		instanceType, ok := node.Labels[corev1.LabelInstanceTypeStable]
+		if !ok {
+			return
+		}
+
+		instanceTypeInfo, ok := instanceTypeInfoMap[instanceType]
+		if !ok || instanceTypeInfo == nil {
+			return
+		}
+
+		// Calculate memory overhead
+		reportedMiB := instanceTypeInfo.MemoryInfo.SizeInMiB
+		actualMiB := node.Status.Capacity.Memory().Value() / (1024 * 1024)
+		overhead := *reportedMiB - actualMiB
```

Comment:
Rather than computing and caching the memory overhead, wouldn't it be simpler to just store the memory directly? I would model this as a capacity override, rather than an override for VM_MEMORY_OVERHEAD_PERCENT.

Reply:
Updated